llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -8284,6 +8284,50 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     visitVectorExtractLastActive(I, Intrinsic);
     return;
   }
+  case Intrinsic::experimental_get_alias_lane_mask: {
+    SDValue SourceValue = getValue(I.getOperand(0));
+    SDValue SinkValue = getValue(I.getOperand(1));
+    SDValue EltSize = getValue(I.getOperand(2));
+    bool IsWriteAfterRead = cast<ConstantSDNode>(getValue(I.getOperand(3)))->getZExtValue() != 0;
+    auto IntrinsicVT = EVT::getEVT(I.getType());
+    auto PtrVT = SourceValue->getValueType(0);
+
+    if (!TLI.shouldExpandGetAliasLaneMask(IntrinsicVT, PtrVT, cast<ConstantSDNode>(EltSize)->getSExtValue())) {
+      visitTargetIntrinsic(I, Intrinsic);
+      return;
+    }
+
+    SDValue Diff = DAG.getNode(ISD::SUB, sdl,
+                              PtrVT, SinkValue, SourceValue);
+    if (!IsWriteAfterRead)
+      Diff = DAG.getNode(ISD::ABS, sdl, PtrVT, Diff);
+
+    Diff = DAG.getNode(ISD::SDIV, sdl, PtrVT, Diff, EltSize);
```

Comment:
Sorry if I'm wrong but wouldn't this line always be executed, even if the `!IsWriteAfterRead` condition is met. So if the `if` statement above is entered, `Diff` is set to the `ABS` node and is then overwritten and set to the `SDIV` node? 
Maybe you forgot a return in the if statement?

Reply:
From my understanding: `!IsWriteAfterRead` would imply that `Diff` is negative, so this inserts the `ISD::ABS` immediately before then using the new `Diff` as an operand to the `SDIV` node, ensuring that it is positive.

Arguably the `if` isn't needed there, as `ABS`-ing a positive number just returns the same number, but then we're adding nodes that we know don't do anything.

Reply:
Oh yes sorry I didn't notice the Diff as an argument in the second assignment.